File: backend/app/modules/question_pipeline/pipeline.py
"""Question pipeline: IR-JSON in, structured questions out.

Independent module — it knows nothing about answers. Deterministic parsing first;
ambiguity routing (and only that) may reach a model, via `graphs/question_graph.py`.
"""
from __future__ import annotations

import re
from dataclasses import dataclass

from app.core.config import settings
from app.modules.question_pipeline.labels import ParsedLabel, parent_of, parse_label, sort_key
from app.schemas.common import Region, union_all
from app.schemas.ir import IRBlock, IRDocument
from app.schemas.pipeline import ExtractedQuestion, QuestionPipelineResult

MARKS_PATTERNS = [
    re.compile(r"[\[\(]\s*(\d{1,3})\s*(?:marks?|m)\s*[\]\)]", re.IGNORECASE),
    re.compile(r"\b(\d{1,3})\s*marks?\b", re.IGNORECASE),
    re.compile(r"[\[\(]\s*(\d{1,3})\s*[\]\)]\s*$"),
]

OPTIONAL_PATTERNS = [
    re.compile(r"\battempt\s+any\b", re.IGNORECASE),
    re.compile(r"\banswer\s+any\b", re.IGNORECASE),
    re.compile(r"\beither\b.*\bor\b", re.IGNORECASE),
    re.compile(r"\bor\b\s*$", re.IGNORECASE),
    re.compile(r"\boptional\b", re.IGNORECASE),
]


@dataclass
class _Candidate:
    label: ParsedLabel
    page: int
    block: IRBlock
    position: int


def extract_questions(document: IRDocument) -> QuestionPipelineResult:
    ordered = document.ordered_blocks()
    candidates = _detect_numbering(ordered)
    if not candidates:
        return QuestionPipelineResult(
            questions=[], ambiguities=["no_numbering_detected"],
            orphan_block_ids=[block.block_id for _, block in ordered],
        )

    questions = _assign_bodies(ordered, candidates)
    questions = _apply_hierarchy(questions)
    orphans = _orphan_blocks(ordered, candidates, questions)
    ambiguities = detect_ambiguities(questions)
    return QuestionPipelineResult(
        questions=questions, ambiguities=ambiguities, orphan_block_ids=orphans
    )


# --------------------------------------------------------------------- stage 1: numbers
def _detect_numbering(ordered: list[tuple[int, IRBlock]]) -> list[_Candidate]:
    candidates: list[_Candidate] = []
    for position, (page, block) in enumerate(ordered):
        parsed = parse_label(block.text, allow_answer_prefix=False)
        if parsed is None:
            continue
        # A label with nothing after it, on a block that is otherwise long prose, is
        # more likely a citation than a question head; require a short prefix.
        if len(parsed.display) > 12:
            continue
        candidates.append(_Candidate(label=parsed, page=page, block=block, position=position))
    return candidates


# ---------------------------------------------------------------- stage 3: body assignment
def _assign_bodies(
    ordered: list[tuple[int, IRBlock]], candidates: list[_Candidate]
) -> list[ExtractedQuestion]:
    questions: list[ExtractedQuestion] = []
    for index, candidate in enumerate(candidates):
        start = candidate.position
        end = candidates[index + 1].position if index + 1 < len(candidates) else len(ordered)
        span = ordered[start:end]

        head_text = candidate.label.remainder
        body_parts = [head_text] if head_text else []
        block_ids = [candidate.block.block_id]
        for _, block in span[1:]:
            body_parts.append(block.text)
            block_ids.append(block.block_id)
        text = " ".join(part for part in body_parts if part).strip()

        regions = _regions_for(span)
        confidences = [block.confidence for _, block in span] or [1.0]
        questions.append(
            ExtractedQuestion(
                question_id=f"q-{candidate.label.normalized}-{index}",
                display_number=candidate.label.display.rstrip(" .:"),
                normalized_number=candidate.label.normalized,
                text=text,
                pages=sorted({page for page, _ in span}),
                regions=regions,
                order_index=index,
                optional=_is_optional(text),
                max_marks=_detect_marks(text),
                confidence=round(min(1.0, sum(confidences) / len(confidences)), 4),
                block_ids=block_ids,
            )
        )
    return questions


def _regions_for(span: list[tuple[int, IRBlock]]) -> list[Region]:
    """One region per page — a question crossing a page boundary gets a region on each."""
    by_page: dict[int, list] = {}
    for page, block in span:
        by_page.setdefault(page, []).append(block.bbox)
    return [Region(page=page, bbox=union_all(boxes)) for page, boxes in sorted(by_page.items())]


# ------------------------------------------------------------------- stage 2: hierarchy
def _apply_hierarchy(questions: list[ExtractedQuestion]) -> list[ExtractedQuestion]:
    """`11(a)` and `11(b)` are separate rows sharing a parent of `11`; `(i)` nests deeper.

    Bare `(a)` / `(i)` labels inherit the nearest preceding numeric stem, so a paper that
    prints `11.` then `(a)` on the next line still produces `11.a`.
    """
    out: list[ExtractedQuestion] = []
    stem_top: str | None = None
    stem_sub: str | None = None
    known = set()
    for question in questions:
        parts = question.normalized_number.split(".")
        if parts[0].isdigit():
            stem_top = parts[0]
            stem_sub = parts[1] if len(parts) > 1 else None
            normalized = question.normalized_number
        elif stem_top is not None:
            is_roman = all(char.isalpha() and char in "ivx" for char in parts[0]) and len(parts[0]) > 0
            if is_roman and stem_sub:
                normalized = f"{stem_top}.{stem_sub}.{parts[0]}"
            else:
                normalized = f"{stem_top}.{parts[0]}"
                stem_sub = parts[0]
        else:
            normalized = question.normalized_number
        while normalized in known:  # duplicate labels must not collide silently
            normalized = f"{normalized}'"
        known.add(normalized)
        parent = parent_of(normalized)
        out.append(
            question.model_copy(
                update={
                    "normalized_number": normalized,
                    "parent_number": parent,
                    "question_id": f"q-{normalized}",
                }
            )
        )
    return out


def _orphan_blocks(
    ordered: list[tuple[int, IRBlock]],
    candidates: list[_Candidate],
    questions: list[ExtractedQuestion],
) -> list[str]:
    if not candidates:
        return [block.block_id for _, block in ordered]
    first = candidates[0].position
    return [block.block_id for _, block in ordered[:first]]


# ------------------------------------------------------------ stage 4: ambiguity checks
def detect_ambiguities(questions: list[ExtractedQuestion]) -> list[str]:
    """Deterministic checks that decide whether a model is needed at all."""
    issues: list[str] = []
    tops = [
        int(question.normalized_number.split(".")[0])
        for question in questions
        if question.normalized_number.split(".")[0].isdigit()
    ]
    unique_tops = sorted(set(tops))
    if tops != sorted(tops):
        issues.append("non_monotonic_numbering")
    for previous, current in zip(unique_tops, unique_tops[1:]):
        if current - previous > 1:
            issues.append(f"gap_between_{previous}_and_{current}")
    for question in questions:
        if not question.text.strip():
            issues.append(f"empty_body:{question.normalized_number}")
        if question.confidence < settings.question_confidence_threshold:
            issues.append(f"low_confidence:{question.normalized_number}")
    return issues


# ----------------------------------------------------------------- stages 5 & 6: extras
def _is_optional(text: str) -> bool:
    return any(pattern.search(text) for pattern in OPTIONAL_PATTERNS)


def _detect_marks(text: str) -> float | None:
    for pattern in MARKS_PATTERNS:
        match = pattern.search(text)
        if match:
            return float(match.group(1))
    return None


def sort_questions(questions: list[ExtractedQuestion]) -> list[ExtractedQuestion]:
    return sorted(questions, key=lambda q: sort_key(q.normalized_number))

File: apps/web/lib/server/modules/question_pipeline/pipeline.ts
/**
 * Question pipeline: IR-JSON in, structured questions out. Port of
 * backend/app/modules/question_pipeline/pipeline.py.
 *
 * Independent module — it knows nothing about answers. Deterministic parsing
 * first; ambiguity routing (and only that) may reach a model, via
 * `graphs/question_graph.ts`.
 */
import { settings } from "@/lib/server/config";
import type { IRBlock, IRDocument } from "../extraction/types";
import { orderedBlocks } from "../extraction/types";
import type { Region } from "../common";
import { unionAll } from "../common";
import { parentOf, parseLabel, type ParsedLabel } from "./labels";
import type { ExtractedQuestion, QuestionPipelineResult } from "./types";

const MARKS_PATTERNS: RegExp[] = [
  /[[(]\s*(\d{1,3})\s*(?:marks?|m)\s*[\])]/i,
  /\b(\d{1,3})\s*marks?\b/i,
  /[[(]\s*(\d{1,3})\s*[\])]\s*$/,
];

// Blocks that are page furniture rather than question text. A printed paper puts its
// marks in a right-hand column and its section headings on their own line; both land
// inside the preceding question's block span. Left in, they corrupt the body text,
// stretch the question's region across the full page width, and a trailing section
// heading pushes the marks marker off the end of the string where MARKS_PATTERNS can
// no longer see it.
const MARKS_ONLY = /^\s*[[(]\s*(\d{1,3})\s*[\])]\s*$/;
const SECTION_HEADER = /^\s*sections?\b.{0,8}$/i;

const OPTIONAL_PATTERNS: RegExp[] = [
  /\battempt\s+any\b/i,
  /\banswer\s+any\b/i,
  /\beither\b.*\bor\b/i,
  /\bor\b\s*$/i,
  /\boptional\b/i,
];

interface Candidate {
  label: ParsedLabel;
  page: number;
  block: IRBlock;
  position: number;
}

export function extractQuestions(document: IRDocument): QuestionPipelineResult {
  const ordered = orderedBlocks(document);
  const candidates = detectNumbering(ordered);
  if (candidates.length === 0) {
    return {
      questions: [],
      ambiguities: ["no_numbering_detected"],
      orphanBlockIds: ordered.map((o) => o.block.blockId),
      usedLlm: false,
    };
  }

  let questions = assignBodies(ordered, candidates);
  questions = applyHierarchy(questions);
  const orphans = orphanBlocks(ordered, candidates);
  const ambiguities = detectAmbiguities(questions);
  return { questions, ambiguities, orphanBlockIds: orphans, usedLlm: false };
}

// --------------------------------------------------------------------- stage 1: numbers
function detectNumbering(ordered: Array<{ page: number; block: IRBlock }>): Candidate[] {
  const candidates: Candidate[] = [];
  ordered.forEach(({ page, block }, position) => {
    const parsed = parseLabel(block.text, false);
    if (parsed === null) return;
    // A label with nothing after it, on a block that is otherwise long prose, is more
    // likely a citation than a question head; require a short prefix.
    if (parsed.display.length > 12) return;
    candidates.push({ label: parsed, page, block, position });
  });
  return candidates;
}

// ---------------------------------------------------------------- stage 3: body assignment
function assignBodies(
  ordered: Array<{ page: number; block: IRBlock }>,
  candidates: Candidate[]
): ExtractedQuestion[] {
  const questions: ExtractedQuestion[] = [];
  candidates.forEach((candidate, index) => {
    const start = candidate.position;
    const end = index + 1 < candidates.length ? candidates[index + 1].position : ordered.length;
    const span = ordered.slice(start, end);

    const headText = candidate.label.remainder;
    const bodyParts: string[] = headText ? [headText] : [];
    const blockIds: string[] = [candidate.block.blockId];
    const content: Array<{ page: number; block: IRBlock }> = [span[0]];
    let marksFromColumn: number | null = null;

    for (const { page, block } of span.slice(1)) {
      const marksMarker = block.text.match(MARKS_ONLY);
      if (marksMarker) {
        marksFromColumn = parseFloat(marksMarker[1]);
        continue;
      }
      if (SECTION_HEADER.test(block.text)) continue;
      bodyParts.push(block.text);
      blockIds.push(block.blockId);
      content.push({ page, block });
    }
    const text = bodyParts.filter(Boolean).join(" ").trim();

    const regions = regionsFor(content);
    const confidences = content.map(({ block }) => block.confidence);
    if (confidences.length === 0) confidences.push(1.0);
    const avgConfidence = confidences.reduce((a, b) => a + b, 0) / confidences.length;

    questions.push({
      questionId: `q-${candidate.label.normalized}-${index}`,
      displayNumber: candidate.label.display.replace(/[\s.:]+$/, ""),
      normalizedNumber: candidate.label.normalized,
      parentNumber: null,
      text,
      pages: [...new Set(content.map((c) => c.page))].sort((a, b) => a - b),
      regions,
      orderIndex: index,
      optional: isOptional(text),
      maxMarks: detectMarks(text) ?? marksFromColumn,
      confidence: Math.round(Math.min(1, avgConfidence) * 10000) / 10000,
      blockIds,
    });
  });
  return questions;
}

/** One region per page — a question crossing a page boundary gets a region on each. */
function regionsFor(span: Array<{ page: number; block: IRBlock }>): Region[] {
  const byPage = new Map<number, IRBlock["bbox"][]>();
  for (const { page, block } of span) {
    const list = byPage.get(page) ?? [];
    list.push(block.bbox);
    byPage.set(page, list);
  }
  return [...byPage.entries()]
    .sort(([a], [b]) => a - b)
    .map(([page, boxes]) => ({ page, bbox: unionAll(boxes) }));
}

// ------------------------------------------------------------------- stage 2: hierarchy
function rstripChars(value: string, chars: string): string {
  let end = value.length;
  while (end > 0 && chars.includes(value[end - 1])) end--;
  return value.slice(0, end);
}

/**
 * `11(a)` and `11(b)` are separate rows sharing a parent of `11`; `(i)` nests deeper.
 * Bare `(a)` / `(i)` labels inherit the nearest preceding numeric stem, so a paper
 * that prints `11.` then `(a)` on the next line still produces `11.a`.
 */
function applyHierarchy(questions: ExtractedQuestion[]): ExtractedQuestion[] {
  const out: ExtractedQuestion[] = [];
  let stemTop: string | null = null;
  let stemSub: string | null = null;
  const known = new Set<string>();

  for (const question of questions) {
    const parts = question.normalizedNumber.split(".");
    let normalized: string;
    if (/^\d+$/.test(parts[0])) {
      stemTop = parts[0];
      stemSub = parts.length > 1 ? parts[1] : null;
      normalized = question.normalizedNumber;
    } else if (stemTop !== null) {
      const isRoman = parts[0].length > 0 && [...parts[0]].every((ch) => "ivx".includes(ch));
      if (isRoman && stemSub) {
        normalized = [stemTop, stemSub, ...parts].join(".");
      } else {
        // `a) i)` arrives as `a.i`: the letter re-bases the sub stem and every deeper
        // level is kept. Taking parts[0] alone silently drops the roman.
        normalized = [stemTop, ...parts].join(".");
        stemSub = parts[0];
      }
    } else {
      normalized = question.normalizedNumber;
    }

    while (known.has(normalized)) {
      // U7 — use a numeric suffix, not an apostrophe.
      const base = rstripChars(rstripChars(normalized, "0123456789"), ".");
      let suffixNum = 2;
      while (known.has(`${base}.${suffixNum}`)) suffixNum++;
      normalized = `${base}.${suffixNum}`;
    }
    known.add(normalized);

    out.push({
      ...question,
      normalizedNumber: normalized,
      parentNumber: parentOf(normalized),
      questionId: `q-${normalized}`,
    });
  }
  return out;
}

function orphanBlocks(
  ordered: Array<{ page: number; block: IRBlock }>,
  candidates: Candidate[]
): string[] {
  if (candidates.length === 0) return ordered.map((o) => o.block.blockId);
  const first = candidates[0].position;
  return ordered.slice(0, first).map((o) => o.block.blockId);
}

// ------------------------------------------------------------ stage 4: ambiguity checks
/** Deterministic checks that decide whether a model is needed at all. */
export function detectAmbiguities(questions: ExtractedQuestion[]): string[] {
  const issues: string[] = [];
  const tops = questions
    .map((q) => q.normalizedNumber.split(".")[0])
    .filter((part) => /^\d+$/.test(part))
    .map((part) => parseInt(part, 10));
  const uniqueTops = [...new Set(tops)].sort((a, b) => a - b);

  const sortedTops = [...tops].sort((a, b) => a - b);
  if (!tops.every((v, i) => v === sortedTops[i])) {
    issues.push("non_monotonic_numbering");
  }
  for (let i = 0; i < uniqueTops.length - 1; i++) {
    const previous = uniqueTops[i];
    const current = uniqueTops[i + 1];
    if (current - previous > 1) issues.push(`gap_between_${previous}_and_${current}`);
  }

  // A parent that prints only `Q2.` and leaves the wording to `a)` / `b)` is the normal
  // shape of a sectioned paper, not an ambiguity worth spending a model call on. Match
  // on prefix, not on parentNumber: a paper that prints `Q12.` then `b) i)` produces
  // `12.b.i` with no intervening `12.b` row, so `12` is an ancestor without ever being
  // anyone's direct parent.
  const numbers = questions.map((q) => q.normalizedNumber);
  for (const question of questions) {
    const stem = `${question.normalizedNumber}.`;
    const hasChildren = numbers.some((n) => n.startsWith(stem));
    if (!question.text.trim() && !hasChildren) {
      issues.push(`empty_body:${question.normalizedNumber}`);
    }
    if (question.confidence < settings.questionConfidenceThreshold) {
      issues.push(`low_confidence:${question.normalizedNumber}`);
    }
  }
  return issues;
}

// ----------------------------------------------------------------- stages 5 & 6: extras
function isOptional(text: string): boolean {
  return OPTIONAL_PATTERNS.some((pattern) => pattern.test(text));
}

function detectMarks(text: string): number | null {
  for (const pattern of MARKS_PATTERNS) {
    const match = text.match(pattern);
    if (match) return parseFloat(match[1]);
  }
  return null;
}
